### DISCUSSION TITLE
Getting final asset filename from addURLDependency

### DISCUSSION THREAD

**Discussion by RamIdeas** [parcel-bundler/parcel#5393] [Q&A] [answered]
I'm trying to upgrade a plugin which handles a specific type of json file from v1 to v2

```
async transform({ asset, ast }) {
    ast.x = asset.addURLDependency(ast.x)
```

that sets ast.x to a 32 character hex string like "36765c662b5bb0c341f4164e30b4fdb2" but I expected it to be the final filename w/ hash (fwiw, the dependency _is_ added to the asset graph)

do I need to do something in the `async generate()` hook to convert this hex string into the final filename?

More complete code examples (both have the same output):

```js
const { Transformer } = require('@parcel/plugin');

module.exports = new Transformer({

	async transform({ asset, ast, config, logger, resolve, options }) {
		ast = JSON.parse(await asset.getCode());
		
		// collect dependencies using asset.addURLDependency()
		collectDependencies(asset, ast);

		asset.type = 'json';
		asset.setCode(JSON.stringify(ast));

		return [asset];
	},
});
```

```js
const { Transformer } = require('@parcel/plugin');

module.exports = new Transformer({
	async parse({ asset, config, logger, resolve, options }) {
		const ast = JSON.parse(await asset.getCode());

		return ast;
	},

	async transform({ asset, ast, config, logger, resolve, options }) {
		// collect dependencies using asset.addURLDependency()
		collectDependencies(asset, ast);

		asset.setAST(ast);

		return [asset];
	},

	async generate({ asset, ast, resolve, options }) {
		const json = options.minify ? JSON.stringify(ast) : JSON.stringify(ast, null, 2);

		return { content: json };
	},
});
```

**Comment by mischnic** [comment, marked as answer]
This is correct. This hash will be replaced with the correct relative url (which depends on the bundle this asset is in and is therefore not yet in known in the transformer) in the packager, e.g.:

https://github.com/parcel-bundler/parcel/blob/064ba0a166ea63ccd2b623dc11324d4cd7ffb526/packages/packagers/raw-url/src/RawUrlPackager.js#L15

**Reply by RamIdeas** [reply]
Thanks for replying!

That's exactly what I expected which is why I was initially stringifying the object during `async generate()` but even that left the hex string in the final output

Are you suggesting that I need to also implement some sort of `replaceURLReferences` in one of the hooks?

**Reply by mischnic** [reply]
> that left the hex string in the final output

Are you using packager that calls `replaceURLReferences `?

> in one of the hooks?

As I've said, this needs to be done in a packager.

Can you provide more details about what your transformer does and especially what packager you're using?

**Reply by RamIdeas** [reply]
I haven't explicitly added a packager. My .parcelrc simply extends the default one and adds my transformer.

Do I need to write a custom packager to do this? Or is there a official one I can add to my .parcelrc?

I'll write a packager that calls `replaceURLReferences` from `@parcel/utils` now and report back -- if it doesn't work I'll follow up with more details answering your last question

**Reply by RamIdeas** [reply]
Ok here's more details:

I have a json file which references local html fragment files.

I am writing a Transformer which knows about the structure of the json file and knows which deeply nested properties refer to those local html files.

For example, assume my json is: `{ "header": "./header.html", "body": "./body.html" }`

I want this json file to be turned into: `{ "header": "header.[hash].html", "body": "body.[hash].html" }` and for those html files to be added to the asset graph and processed like any other html file

My config looks like:

```
{
	"extends": "@parcel/config-default",
	"transformers": {
		"*.xyz.json": ["@my-scope/parcel-transformer-xyz-json"]
	}
}
```

**Reply by RamIdeas** [reply]
Ah I got it working by adding the raw-url packager to my config

```json
{
	"extends": "@parcel/config-default",
	"transformers": {
		"*.xyz.json": ["@my-scope/parcel-transformer-xyz-json"]
	},
	"packagers": {
		"*.xyz.json": "@parcel/packager-raw-url"
	}
}
```

Thanks for your help!
